Various Improvements / Additions

Sinuousity opened the pull request:
- Expanded the implementation of basic object interaction.
- Implemented a new system for object materials that procedurally creates them from object data rather than using preset materials. Also extended this by adding multiple texture map support to materials via new shaders.

Sinuousity committed:
Added fix for interact text creating slow down.

Sinuousity committed:
Improved/Changed the handling of material conversion / creation. Added 3 shaders for the types of materials. Expanded alphaProperty within the MWMaterialProperties class to read directly into material (However much of the data seems to not be used). Maintained material cache. Adapted NIFObjectBuilder to use the new arrangement.

Sinuousity committed:
Added a toggle to disable the MusicPlayer on scene start. MP3 streaming has a large impact on performance at certain intervals and should probably be avoided without a better more performant solution in the future.

Sinuousity committed:
Added simple mathematics based animation to interior cell doors. Added GameObject active state conditionals in cell detail radius check.

Sinuousity committed:
Added exterior cell sun shadows toggle. Cleanup

Sinuousity committed:
Began adding support for multiple texture types per material ( Occlusion / Detail / Gloss / Glow / Bump )
Minor syntactical fixes. Clearing up debug log clutter.

Sinuousity committed:
Extended multiple material maps support.

Sinuousity committed:
Added container interact text support. Container naming needs formatting.

Sinuousity committed:
Added cursor locking.
Tilde/Backquote to free the cursor lock.

Sinuousity committed:
Added BOOK and MISC Record support for interact text. Fixed CONT Record names not being set correctly on container components.

Sinuousity committed:
Added early WEAP and INGR Record support to interaction.
Cleanup

Sinuousity committed:
Meta files push

Sinuousity committed:
Added Book TEXT dump to log.
F key while targeting a book object.

ColeDeanShepherd commented:
Wow. That is a lot of great progress! I'm going to finish reviewing it all tomorrow before I merge it.

ColeDeanShepherd commented on Assets/Scripts/TES/MorrowindEngine.cs:
```diff
@@ -285,33 +286,33 @@ public void Update()
-
 			// Cast a ray to see what the camera is looking at.
-			var ray = new Ray(playerCameraObj.transform.position, playerCameraObj.transform.forward);
+			Ray ray = new Ray(playerCameraObj.transform.position, playerCameraObj.transform.forward);
```
I generally use "var" if the type is clear from the initial value. This is a minor style preference, so I don't mind if others prefer to avoid "var", but please don't Ctrl+F all the "var"s and replace them with types! They can cut down on verbosity sometimes.

ColeDeanShepherd commented on Assets/Scripts/TES/MorrowindEngine.cs:
```diff
@@ -285,33 +286,33 @@ public void Update()
+					DoorComponent doorComponent = doorObj.GetComponent<DoorComponent>();
 
-					interactTextObj.SetActive(true);
+					if ( !interactTextObj.activeSelf ) interactTextObj.SetActive( true );
```
Try to stay consistent in coding style with the rest of the code. I don't put spaces around parenthesis, just another minor style choice. Once again, it's not a huge deal, I'll merge it, but I'm going to try to clean up a lot of code soon and I might change it back. It'll just become a mess if different contributors use different style conventions everywhere.

Sinuousity replied:
Noted! the var thing was most likely me trying to follow the code better. having never seen the classes and types that this engine uses, I'm still learning the structure. I'll try to use var in the same way you do, and I'll change my formatting settings in Visual Studio to match yours as closely as I can.

ColeDeanShepherd replied:
I think there's a way to change formatting settings per-project so that you don't mess up the formatting in your other projects. I haven't looked into that though, so I don't know how.

ColeDeanShepherd commented on Assets/Scripts/TES/MorrowindEngine.cs:
```diff
@@ -854,7 +860,7 @@ private GameObject CreatePlayerCamera(Vector3 position)
 			var camera = GameObjectUtils.CreateMainCamera(position, Quaternion.identity);
 			camera.GetComponent<Camera>().cullingMask = ~(1 << markerLayer);
-
+			Camera.main.renderingPath = RenderingPath.DeferredShading;
```
My computer is terrible, so I avoided using deferred shading. At the very least I'd like to add an option to enable/disable deferred rendering in the inspector, or better, in the main menu. So I might remove this line from the request later and add an option for deferred rendering in another commit.

Sinuousity replied:
Of course! i had actually meant to make it a toggle on the TESUnity object but hadnt gotten around to it yet. that would definitely be optimal.

ColeDeanShepherd commented on Assets/Scripts/Audio/AudioUtils.cs:
```diff
@@ -50,7 +50,7 @@ public static GameObject Play2DAudioStream(MP3StreamReader audioStream)
 		var audioSource = gameObject.AddComponent<AudioSource>();
 		//audioSource.clip = CreateStreamingAudioClip("tmpAudioClip", audioStream);
-		//audioSource.loop = true;
+		audioSource.loop = true;
```
So, long story short, I ended up implementing a hack to get runtime MP3 streaming working. I'm using OnAudioFilterRead to play the music right now, and the only reason I'm adding an AudioSource is because that enables OnAudioFilterRead IIRC. So setting audioSource.loop = true here won't have any effect. It will have an effect when I fix the audio hacks, though, so I'll leave this change in. Just FYI.

Sinuousity replied:
Oh sorry, I actually didn't mean to leave that one in, I was just making some of the debug log warnings go away and forgot to change that one back.

ColeDeanShepherd commented on Assets/Scripts/TES/PlayerComponent.cs:
```diff
@@ -88,6 +88,22 @@ private void FixedUpdate()
 
 		private void Rotate()
 		{
+			if ( Cursor.lockState != CursorLockMode.Locked )
```
Leaving a note for myself to add these controls to the README.

ColeDeanShepherd replied:
Nevermind, you already did.

ColeDeanShepherd commented:
Alright, I reviewed the changes and they look great! Thanks for all your work Sinuousity. Before I merge your request I want to check out and test your changes locally. It's a little late so I'll do that in the morning.

Sinuousity committed:
Fix for transparent material shadows. Alpha blended material + alpha tested shadows

Sinuousity committed:
Changes to the book object component. New structure for components might be more efficient, but needs revisal and tweaking.

Sinuousity committed:
Hefty changes to the way that the object component system works. Now, instead of having separate component classes for each type of object, all objects share one component that encompasses functionality.
This unification made it much easier to add support for the remaining Record types, which I've done.
